lib/utils.js

```diff
@@ -1,246 +1,212 @@
+    }
   }
 
-  function request(first) {
-    if (first) opts.url = setPageLimit(opts.url);
+  const promise = paged(opts.url, opts, nextPage)
+    .then(res => result(res, options, opts, options));
 
-    utils.request(opts, function(err, data, stream) {
-      if (err) {
-        cb(err);
-        return;
-      }
-
-      try {
-        if (data && Array.isArray(data)) {
-          arr.push.apply(arr, data);
-        }
-        var page = utils.parseLink(stream.headers.link);
-        if (page && page.next) {
-          opts.url = page.next.url;
-          request();
-        } else {
-          cb(null, arr, stream);
-        }
-      } catch (err) {
-        cb(err);
-      }
-    });
+  if (typeof callback === 'function') {
+    return promise.then(res => callback(null, res)).catch(callback);
   }
-  request(true);
+
+  return promise;
 };
 
 /**
- * Default settings to use for GitHub API requests.
+ * Normalize request options object, request body, headers and related properties
+ * to use for GitHub API requests.
  *
- * @param  {Object} `options`
- * @return {Object}
+ * @param  {Object} `appOptions` Options on the instance of github-base.
+ * @return {Function} Returns a function that takes a request method name, the URL to expand, and method options.
  */
 
-utils.defaults = function(options) {
-  return function(method, url, data) {
-    var opts = utils.extend({method: 'get'}, options);
-    opts.apiurl = opts.apiurl || 'https://api.github.com';
-    var config = lowercaseKeys(opts);
-    opts = utils.merge({}, config, data);
-    opts = interpolate(url, opts);
-    opts = createURL(opts);
-    opts = lowercaseKeys(opts);
-    opts = createAuth(opts);
-    opts = cleanup(opts);
-    opts.method = method.toLowerCase();
-    opts = normalizeOpts(opts, config);
-
-    var defaults = {
-      'accept': 'application/json',
-      'user-agent': 'https://github.com/jonschlinkert/github-base'
-    };
-
-    opts.headers = utils.extend({}, defaults, lowercaseKeys(opts.headers));
-    return opts;
-  };
-};
+exports.defaults = function(appOptions) {
+  const defaults = { apiurl: 'https://api.github.com', method: 'get', params: [] };
 
-/**
- * Replace params with actual values.
- */
+  return function(method, uri, methodOptions) {
+    const options = Object.assign({}, defaults, appOptions, methodOptions, { uri, method });
+    const context = { orig: clone(options), options: lowercaseKeys(options) };
+    context.options.url = interpolate(uri, context);
+    context.options.headers = createHeaders(context.options);
 
-function interpolate(path, options) {
-  var opts = utils.extend({}, options, {params: []});
-  opts.url = path.replace(/:([\w_]+)/g, function(m, prop) {
-    opts.params.push(prop);
-    return opts[prop] || prop;
-  });
-  return opts;
-}
+    let opts = context.options;
+    const body = omit(context.orig, Object.keys(opts));
+    const json = opts.json;
+    const text = opts.body;
```

Comment:
Something doesn't seem right about `text` here. It's assigned the value of `opts.body` here, then later if `text` has a value it's assigned back to `opts.body`. What doesn't seem right is that `text` and `opts.body` aren't modified in any way between these points.

Reply:
Since `opts = omit(opts, bodyKeys);` omits `body` keys, `if (text)` re-adds `opts.body` if it was omitted.

Reply:
Thanks, that makes sense now.